File: go/core/network_accountant_test.go
package core

import (
	"fmt"
	"testing"
)

// boolPtr is a tiny test helper for nullable bool literals (Go has no
// nullable-bool syntax). Mirrors the adapters/netbytes_test.go helper —
// duplicated here because this file lives in the `core` package.
func boolPtr(b bool) *bool { return &b }

// --- Basic record + finalize -----------------------------------------------

func TestNetworkAccountant_RecordUpdatesCounters(t *testing.T) {
	a := NewNetworkAccountant()
	a.Record("a.com", 100, 10, nil)
	a.Record("a.com", 50, 5, nil)
	snap := a.Finalize()
	if snap.BytesIn != 150 {
		t.Fatalf("BytesIn = %d, want 150", snap.BytesIn)
	}
	if snap.BytesOut != 15 {
		t.Fatalf("BytesOut = %d, want 15", snap.BytesOut)
	}
	if snap.CallCount != 2 {
		t.Fatalf("CallCount = %d, want 2", snap.CallCount)
	}
}

func TestNetworkAccountant_FinalizeGroupsByHost(t *testing.T) {
	a := NewNetworkAccountant()
	a.Record("a.com", 100, 10, nil)
	a.Record("b.com", 200, 20, nil)
	hosts := a.Finalize().ByHost["hosts"].([]map[string]interface{})
	byName := map[string]map[string]interface{}{}
	for _, h := range hosts {
		byName[h["host"].(string)] = h
	}
	got := byName["a.com"]
	if got["calls"].(int64) != 1 {
		t.Fatalf("a.com calls = %v, want 1", got["calls"])
	}
	if got["bytes_in"].(int64) != 100 {
		t.Fatalf("a.com bytes_in = %v, want 100", got["bytes_in"])
	}
	if got["bytes_out"].(int64) != 10 {
		t.Fatalf("a.com bytes_out = %v, want 10", got["bytes_out"])
	}
	// isInternal=nil → bytes_out attributes as external (v2 §6.1).
	if got["external_bytes_out"].(int64) != 10 {
		t.Fatalf("a.com external_bytes_out = %v, want 10", got["external_bytes_out"])
	}
}

func TestNetworkAccountant_FinalizeCapsToTop20WithOtherBucket(t *testing.T) {
	a := NewNetworkAccountant()
	// 25 hosts; host_i gets i+1 bytes_in so heavy ones are deterministic.
	for i := 0; i < 25; i++ {
		a.Record(fmt.Sprintf("h%02d.com", i), int64(i)+1, 0, nil)
	}
	hosts := a.Finalize().ByHost["hosts"].([]map[string]interface{})
	if len(hosts) != FinalizeCap+1 {
		t.Fatalf("len(hosts) = %d, want %d", len(hosts), FinalizeCap+1)
	}
	names := map[string]bool{}
	for _, h := range hosts {
		names[h["host"].(string)] = true
	}
	if !names["_other"] {
		t.Fatal("_other bucket missing")
	}
	if !names["h24.com"] {
		t.Fatal("heaviest host h24.com missing")
	}
	if names["h00.com"] {
		t.Fatal("lightest host h00.com should be folded into _other")
	}
	var other map[string]interface{}
	for _, h := range hosts {
		if h["host"].(string) == "_other" {
			other = h
			break
		}
	}
	// 5 lightest folded: bytes_in = 1+2+3+4+5 = 15, calls = 5.
	if other["calls"].(int64) != 5 {
		t.Fatalf("_other.calls = %v, want 5", other["calls"])
	}
	if other["bytes_in"].(int64) != 15 {
		t.Fatalf("_other.bytes_in = %v, want 15", other["bytes_in"])
	}
}

func TestNetworkAccountant_EmptyFinalize(t *testing.T) {
	hosts := NewNetworkAccountant().Finalize().ByHost["hosts"].([]map[string]interface{})
	if len(hosts) != 0 {
		t.Fatalf("expected empty hosts, got %d entries", len(hosts))
	}
}

func TestNetworkAccountant_LiveCapFoldsOverflowIntoOther(t *testing.T) {
	a := NewNetworkAccountant()
	for i := 0; i < LiveCap+50; i++ {
		a.Record(fmt.Sprintf("host%d.com", i), 0, 1, boolPtr(false))
	}
	// Before finalize, only LiveCap distinct hosts in the map.
	if a.LiveHostCount() != LiveCap {
		t.Fatalf("LiveHostCount = %d, want %d", a.LiveHostCount(), LiveCap)
	}
	hosts := a.Finalize().ByHost["hosts"].([]map[string]interface{})
	var other map[string]interface{}
	for _, h := range hosts {
		if h["host"].(string) == "_other" {
			other = h
			break
		}
	}
	if other == nil {
		t.Fatal("_other bucket missing")
	}
	// 50 hosts that overflowed LiveCap + (LiveCap - FinalizeCap) folded
	// from the top-N cap = (LiveCap + 50 - FinalizeCap) entries.
	want := int64(LiveCap + 50 - FinalizeCap)
	if other["calls"].(int64) != want {
		t.Fatalf("_other.calls = %v, want %d", other["calls"], want)
	}
}

func TestNetworkAccountant_FrozenAfterFinalizeRecordIsNoop(t *testing.T) {
	a := NewNetworkAccountant()
	a.Record("a.com", 100, 10, nil)
	snap1 := a.Finalize()
	a.Record("b.com", 999, 999, nil)
	snap2 := a.Finalize()
	if snap1.BytesIn != snap2.BytesIn {
		t.Fatalf("BytesIn changed after freeze: %d → %d", snap1.BytesIn, snap2.BytesIn)
	}
	if snap1.CallCount != snap2.CallCount {
		t.Fatalf("CallCount changed after freeze: %d → %d", snap1.CallCount, snap2.CallCount)
	}
}

func TestNetworkAccountant_EmptyHostFallsBackToUnknown(t *testing.T) {
	a := NewNetworkAccountant()
	a.Record("", 10, 0, nil)
	hosts := a.Finalize().ByHost["hosts"].([]map[string]interface{})
	if hosts[0]["host"].(string) != "_unknown" {
		t.Fatalf("empty host got %q, want _unknown", hosts[0]["host"])
	}
}

func TestNetworkAccountant_NegativeBytesClampedToZero(t *testing.T) {
	a := NewNetworkAccountant()
	a.Record("a.com", -10, -20, nil)
	snap := a.Finalize()
	if snap.BytesIn != 0 {
		t.Fatalf("negative bytes_in not clamped: %d", snap.BytesIn)
	}
	if snap.BytesOut != 0 {
		t.Fatalf("negative bytes_out not clamped: %d", snap.BytesOut)
	}
}

func TestNetworkAccountant_SyntheticOtherCollidesWithRealHostNamedOther(t *testing.T) {
	a := NewNetworkAccountant()
	a.Record("_other", 100, 50, nil)
	a.Record("real.com", 1, 1, nil)
	hosts := a.Finalize().ByHost["hosts"].([]map[string]interface{})
	otherCount := 0
	var other map[string]interface{}
	for _, h := range hosts {
		if h["host"].(string) == "_other" {
			otherCount++
			other = h
		}
	}
	if otherCount != 1 {
		t.Fatalf("expected exactly one _other entry, got %d", otherCount)
	}
	if other["bytes_in"].(int64) != 100 {
		t.Fatalf("_other.bytes_in = %v, want 100 (real-host folded in)", other["bytes_in"])
	}
}

// --- External-byte split (v2) ----------------------------------------------

func TestNetworkAccountant_InternalCallDoesNotContributeToExternal(t *testing.T) {
	a := NewNetworkAccountant()
	a.Record("10.0.0.5", 100, 200, boolPtr(true))
	snap := a.Finalize()
	if snap.ExternalBytesOut != 0 {
		t.Fatalf("internal call should not contribute to external: got %d", snap.ExternalBytesOut)
	}
	host := snap.ByHost["hosts"].([]map[string]interface{})[0]
	if host["external_bytes_out"].(int64) != 0 {
		t.Fatalf("per-host external_bytes_out = %v, want 0", host["external_bytes_out"])
	}
	// Raw bytes_out still recorded.
	if host["bytes_out"].(int64) != 200 {
		t.Fatalf("per-host bytes_out = %v, want 200", host["bytes_out"])
	}
}

func TestNetworkAccountant_PublicCallContributesToExternal(t *testing.T) {
	a := NewNetworkAccountant()
	a.Record("api.example.com", 100, 500, boolPtr(false))
	snap := a.Finalize()
	if snap.ExternalBytesOut != 500 {
		t.Fatalf("public call external = %d, want 500", snap.ExternalBytesOut)
	}
}

func TestNetworkAccountant_NilIsInternalTreatedAsExternal(t *testing.T) {
	a := NewNetworkAccountant()
	a.Record("api.example.com", 100, 500, nil)
	snap := a.Finalize()
	if snap.ExternalBytesOut != 500 {
		t.Fatalf("nil isInternal external = %d, want 500", snap.ExternalBytesOut)
	}
}

// Property invariant (v2 §10.3 #1): scalar external == sum of per-host external.
func TestNetworkAccountant_ScalarEqualsSumOfPerHostExternal(t *testing.T) {
	a := NewNetworkAccountant()
	a.Record("a.com", 0, 100, boolPtr(false))
	a.Record("b.com", 0, 200, boolPtr(false))
	a.Record("10.0.0.1", 0, 999, boolPtr(true))
	snap := a.Finalize()
	var sum int64
	for _, h := range snap.ByHost["hosts"].([]map[string]interface{}) {
		sum += h["external_bytes_out"].(int64)
	}
	if sum != snap.ExternalBytesOut {
		t.Fatalf("sum(per-host external) = %d, scalar = %d", sum, snap.ExternalBytesOut)
	}
	if snap.ExternalBytesOut != 300 {
		t.Fatalf("ExternalBytesOut = %d, want 300", snap.ExternalBytesOut)
	}
}

// Mirror python's "_other carries external bytes through LIVE_CAP + top-20 folds".
func TestNetworkAccountant_OtherBucketCarriesExternalBytes(t *testing.T) {
	a := NewNetworkAccountant()
	for i := 0; i < LiveCap; i++ {
		a.Record(fmt.Sprintf("host%d.com", i), 0, 1, boolPtr(false))
	}
	a.Record("overflow.com", 0, 555, boolPtr(false))
	hosts := a.Finalize().ByHost["hosts"].([]map[string]interface{})
	var other map[string]interface{}
	for _, h := range hosts {
		if h["host"].(string) == "_other" {
			other = h
			break
		}
	}
	if other == nil {
		t.Fatal("_other bucket missing")
	}
	// 555 from the overflow + (LiveCap - FinalizeCap) folded from top-20 cap.
	want := int64(LiveCap-FinalizeCap) + 555
	if other["external_bytes_out"].(int64) != want {
		t.Fatalf("_other.external_bytes_out = %v, want %d", other["external_bytes_out"], want)
	}
}

func TestNetworkAccountant_DefaultIsInternalRoutesBytesAsExternal(t *testing.T) {
	a := NewNetworkAccountant()
	a.Record("api.example.com", 0, 100, nil)
	snap := a.Finalize()
	if snap.ExternalBytesOut != 100 {
		t.Fatalf("default nil → external = %d, want 100", snap.ExternalBytesOut)
	}
}

// --- Registry --------------------------------------------------------------

func TestAccountantRegistry_RegisterAndGet(t *testing.T) {
	ResetAccountantRegistryForTests()
	a := NewNetworkAccountant()
	RegisterAccountant("t-1", a)
	got := GetAccountant("t-1")
	if got != a {
		t.Fatalf("GetAccountant returned a different instance: %p vs %p", got, a)
	}
}

func TestAccountantRegistry_GetMissingReturnsNil(t *testing.T) {
	ResetAccountantRegistryForTests()
	if got := GetAccountant("does-not-exist"); got != nil {
		t.Fatalf("expected nil for missing task, got %p", got)
	}
}

func TestAccountantRegistry_UnregisterReturnsThenRemoves(t *testing.T) {
	ResetAccountantRegistryForTests()
	a := NewNetworkAccountant()
	RegisterAccountant("t-1", a)
	got := UnregisterAccountant("t-1")
	if got != a {
		t.Fatalf("Unregister returned a different instance")
	}
	if GetAccountant("t-1") != nil {
		t.Fatal("Unregister did not remove the entry")
	}
	// Idempotent: a second call returns nil.
	if UnregisterAccountant("t-1") != nil {
		t.Fatal("repeated Unregister did not return nil")
	}
}

File: typescript/tests/network-accountant.test.ts
/**
 * NetworkAccountant tests — port of python/tests/test_network_accountant.py
 * and test_network_accountant_external.py.
 */

import { describe, it, expect, beforeEach } from "vitest";
import {
  NetworkAccountant,
  FINALIZE_CAP,
  LIVE_CAP,
  registerAccountant,
  getAccountant,
  unregisterAccountant,
  _resetAccountantRegistryForTests,
} from "../src/adapters/network-accountant.js";

describe("NetworkAccountant — record + finalize", () => {
  it("record updates counters", () => {
    const a = new NetworkAccountant();
    a.record("a.com", 100, 10);
    a.record("a.com", 50, 5);
    const snap = a.finalize();
    expect(snap.bytesIn).toBe(150);
    expect(snap.bytesOut).toBe(15);
    expect(snap.callCount).toBe(2);
  });

  it("finalize groups by host", () => {
    const a = new NetworkAccountant();
    a.record("a.com", 100, 10);
    a.record("b.com", 200, 20);
    const snap = a.finalize();
    const a_host = snap.byHost.hosts.find((h) => h.host === "a.com")!;
    expect(a_host.calls).toBe(1);
    expect(a_host.bytes_in).toBe(100);
    expect(a_host.bytes_out).toBe(10);
    // isInternal default null → bytes_out attributes as external.
    expect(a_host.external_bytes_out).toBe(10);
  });

  it("finalize caps to top 20 with _other bucket", () => {
    const a = new NetworkAccountant();
    for (let i = 0; i < 25; i++) {
      a.record(`h${String(i).padStart(2, "0")}.com`, i + 1, 0);
    }
    const hosts = a.finalize().byHost.hosts;
    expect(hosts.length).toBe(FINALIZE_CAP + 1);
    const names = new Set(hosts.map((h) => h.host));
    expect(names.has("_other")).toBe(true);
    expect(names.has("h24.com")).toBe(true);
    expect(names.has("h00.com")).toBe(false);
    const other = hosts.find((h) => h.host === "_other")!;
    expect(other.calls).toBe(5);
    expect(other.bytes_in).toBe(1 + 2 + 3 + 4 + 5);
  });

  it("empty finalize returns empty hosts array", () => {
    expect(new NetworkAccountant().finalize().byHost).toEqual({ hosts: [] });
  });

  it("LIVE_CAP overflow folds into _other", () => {
    const a = new NetworkAccountant();
    for (let i = 0; i < LIVE_CAP + 50; i++) {
      a.record(`host${i}.com`, 0, 1, false);
    }
    expect(a.liveHostCount()).toBe(LIVE_CAP);
    const hosts = a.finalize().byHost.hosts;
    const other = hosts.find((h) => h.host === "_other")!;
    expect(other.calls).toBe(LIVE_CAP + 50 - FINALIZE_CAP);
  });

  it("frozen after finalize — subsequent record is no-op", () => {
    const a = new NetworkAccountant();
    a.record("a.com", 100, 10);
    const snap1 = a.finalize();
    a.record("b.com", 999, 999);
    const snap2 = a.finalize();
    expect(snap1.bytesIn).toBe(snap2.bytesIn);
    expect(snap1.callCount).toBe(snap2.callCount);
  });

  it("empty host falls back to _unknown", () => {
    const a = new NetworkAccountant();
    a.record("", 10, 0);
    const hosts = a.finalize().byHost.hosts;
    expect(hosts[0].host).toBe("_unknown");
  });

  it("negative bytes clamped to zero", () => {
    const a = new NetworkAccountant();
    a.record("a.com", -10, -20);
    const snap = a.finalize();
    expect(snap.bytesIn).toBe(0);
    expect(snap.bytesOut).toBe(0);
  });

  it("synthetic _other collides with real host literally named '_other'", () => {
    const a = new NetworkAccountant();
    a.record("_other", 100, 50);
    a.record("real.com", 1, 1);
    const hosts = a.finalize().byHost.hosts;
    const otherCount = hosts.filter((h) => h.host === "_other").length;
    expect(otherCount).toBe(1);
    const other = hosts.find((h) => h.host === "_other")!;
    expect(other.bytes_in).toBe(100);
  });
});

describe("NetworkAccountant — external-byte split (v2)", () => {
  it("internal call does not contribute to external", () => {
    const a = new NetworkAccountant();
    a.record("10.0.0.5", 100, 200, true);
    const snap = a.finalize();
    expect(snap.externalBytesOut).toBe(0);
    const host = snap.byHost.hosts[0];
    expect(host.external_bytes_out).toBe(0);
    expect(host.bytes_out).toBe(200);
  });

  it("public call contributes to external", () => {
    const a = new NetworkAccountant();
    a.record("api.example.com", 100, 500, false);
    expect(a.finalize().externalBytesOut).toBe(500);
  });

  it("null isInternal treated as external (conservative)", () => {
    const a = new NetworkAccountant();
    a.record("api.example.com", 100, 500, null);
    expect(a.finalize().externalBytesOut).toBe(500);
  });

  // v2 §10.3 invariant 1: scalar external == sum of per-host external.
  it("scalar external equals sum of per-host external", () => {
    const a = new NetworkAccountant();
    a.record("a.com", 0, 100, false);
    a.record("b.com", 0, 200, false);
    a.record("10.0.0.1", 0, 999, true);
    const snap = a.finalize();
    const sum = snap.byHost.hosts.reduce(
      (acc, h) => acc + (h.external_bytes_out as number),
      0,
    );
    expect(sum).toBe(snap.externalBytesOut);
    expect(snap.externalBytesOut).toBe(300);
  });

  it("_other bucket carries external bytes through LIVE_CAP + top-20 folds", () => {
    const a = new NetworkAccountant();
    for (let i = 0; i < LIVE_CAP; i++) {
      a.record(`host${i}.com`, 0, 1, false);
    }
    a.record("overflow.com", 0, 555, false);
    const hosts = a.finalize().byHost.hosts;
    const other = hosts.find((h) => h.host === "_other")!;
    expect(other.external_bytes_out).toBe(LIVE_CAP - FINALIZE_CAP + 555);
  });

  it("default null isInternal routes bytes as external", () => {
    const a = new NetworkAccountant();
    a.record("api.example.com", 0, 100);
    expect(a.finalize().externalBytesOut).toBe(100);
  });
});

describe("NetworkAccountant — registry", () => {
  beforeEach(() => {
    _resetAccountantRegistryForTests();
  });

  it("register then get returns same instance", () => {
    const a = new NetworkAccountant();
    registerAccountant("t-1", a);
    expect(getAccountant("t-1")).toBe(a);
  });

  it("get missing returns undefined", () => {
    expect(getAccountant("does-not-exist")).toBeUndefined();
  });

  it("unregister returns then removes; idempotent", () => {
    const a = new NetworkAccountant();
    registerAccountant("t-1", a);
    expect(unregisterAccountant("t-1")).toBe(a);
    expect(getAccountant("t-1")).toBeUndefined();
    expect(unregisterAccountant("t-1")).toBeUndefined();
  });
});
